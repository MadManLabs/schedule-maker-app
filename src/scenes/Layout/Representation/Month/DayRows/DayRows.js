import React from 'react';
import PropTypes from 'prop-types'
import { Grid } from 'semantic-ui-react';

const DayRows = ({ displayMonth, displayYear, dateService }) => (

  generateRows(splitDaysIntoWeeks(padDaysInMonthArr(displayYear, displayMonth, dateService)))

);

/**
 * For display on the month grid, if the first day of the month isn't sunday, null's need to be added
 * onto the beggining of the array in order to adjust the display to the right
 * EG if a month starts on wednesday, we want 3 null's padded onto the front of the days array 
 * in order to adjust for sun through tue
 */
function padDaysInMonthArr(year, month, dateService) {
  
  const daysArr = dateService.getDaysInMonth(year, month);
  const dayCount = dateService.getDaysCountInMonth(year, month);
  const initialPaddingNeeded = daysArr[0];
  return Array.from({ length: initialPaddingNeeded }, () => null)
    .concat(Array.from({ length: dayCount }, (_, i) => i + 1));
}

//Transform the 1d padded array into a 2d array (5 x 7) representing the weeks of the month
function splitDaysIntoWeeks(paddedArr) {
  
  const chunckArr = [];
  for (let i = 0; i < 36; i += 7) {
    chunckArr.push(paddedArr.slice(i, i + 7));
  }
  return chunckArr;
}

//Consume the 2d array representing the weeks, and generate  5-6 Grid.Rows with the days as columns 
function generateRows(chunkedArr) {
  
  return chunkedArr.map((subArr, i) => {
    return <Grid.Row key={i}>
      {subArr.map((day, j) => {
        return (<Grid.Column key={j}>
          <div>{day}</div>
        </Grid.Column>
        )
      })}
    </Grid.Row>
  })
}

const propTypes = {
  displayMonth: PropTypes.number.isRequired,
  displayYear: PropTypes.number.isRequired,
  dateService: PropTypes.object.isRequired
};

const defaultProps = {};

DayRows.propTypes = propTypes;

DayRows.defaultProps = defaultProps;

export default DayRows;
