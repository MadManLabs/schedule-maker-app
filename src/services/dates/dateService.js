class DateService {
  constructor() {
    this.today = new Date();
    this.dayStrRepArr = ['Sun', 'Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat'];
    this.monthNamesShort = ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec'];
    this.monthNamesLong = ['January', 'Febuary', 'March', 'April', 'May', 'June', 'July',
      'August', 'September', 'Oct', 'Nov', 'Dec'];
  }
  // getDaysCountInMonth :: (Num, Num) -> Num
  getDaysCountInMonth(year = this.today.getFullYear(), month = this.today.getMonth()) {
    //CF https://stackoverflow.com/a/1184359
    return new Date(year, month + 1, 0).getDate();
  }
  // getDaysInMonth :: (Num, Num) -> [Num]
  getDaysInMonth(year = this.today.getFullYear(), month = this.today.getMonth()) {
    return Array.from({ length: this.getDaysCountInMonth(year, month) }, (_, i) => {
      return new Date(year, month, i + 1).getDay();
    })
  }
  // getNamedDaysInMonth :: (Num, Num) -> [String]
  getNamedDaysInMonth(year = this.today.getFullYear(), month = this.today.getMonth()) {
    return this.getDaysInMonth(year, month).map(dayNumber => this.dayStrRepArr[dayNumber]);
  }
}

const dateService = new DateService();
export default dateService;
