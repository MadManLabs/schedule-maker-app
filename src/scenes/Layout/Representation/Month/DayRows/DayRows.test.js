import { cellColor, cellClasses } from './DayRows';
import React from 'react';
import renderer from 'react-test-renderer';
import dateService from '../../../../.././services/dates/dateService';
import DayRows from './DayRows.js';

test('Cell color should be default if there is a day in it', () => {
    expect(cellColor(1)).toBe(undefined);
});

test('Cell color should be gray if there is NO day in that cell', () => {
    expect(cellColor(undefined)).toBe("grey");
});

test('Cell color should be gray if there is NO day in that cell', () => {
    expect(cellColor(-1)).toBe("grey");
});

test('Cell color should be gray if there is NO day in that cell', () => {
    expect(cellColor("asd")).toBe("grey");
});

test('Cell should have no classes if it has a day in it', () => {
    expect(cellClasses(1)).toBe(undefined);
});

test('Cell classes should be appropriate if it DOES NOT have a day in it', () => {
    expect(cellClasses(undefined)).toBe("no-day shadow");
});

test('Cell classes should be appropriate if it DOES NOT have a day in it', () => {
    expect(cellClasses(-1)).toBe("no-day shadow");
});

test('Cell classes should be appropriate if it DOES NOT have a day in it', () => {
    expect(cellClasses("asd")).toBe("no-day shadow");
});

it('renders Oct 2017 correctly', () => {
    const tree = renderer.create(
        <DayRows displayMonth={9} displayYear={2017} dateService={dateService} />
    ).toJSON();
    expect(tree).toMatchSnapshot();
});
it('renders Feb 2016 correctly', () => {
    const tree = renderer.create(
        <DayRows displayMonth={1} displayYear={2016} dateService={dateService} />
    ).toJSON();
    expect(tree).toMatchSnapshot();
})
