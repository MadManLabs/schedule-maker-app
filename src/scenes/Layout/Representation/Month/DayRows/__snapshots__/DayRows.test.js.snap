// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Feb 2016 correctly 1`] = `
Array [
  <div
    className="row"
>
    <div
        className="column"
    >
        <div>
            1
        </div>
    </div>
    <div
        className="column"
    >
        <div>
            2
        </div>
    </div>
    <div
        className="column"
    >
        <div>
            3
        </div>
    </div>
    <div
        className="column"
    >
        <div>
            4
        </div>
    </div>
    <div
        className="column"
    >
        <div>
            5
        </div>
    </div>
    <div
        className="column"
    >
        <div>
            6
        </div>
    </div>
    <div
        className="column"
    >
        <div>
            7
        </div>
    </div>
</div>,
  <div
    className="row"
>
    <div
        className="column"
    >
        <div>
            8
        </div>
    </div>
    <div
        className="column"
    >
        <div>
            9
        </div>
    </div>
    <div
        className="column"
    >
        <div>
            10
        </div>
    </div>
    <div
        className="column"
    >
        <div>
            11
        </div>
    </div>
    <div
        className="column"
    >
        <div>
            12
        </div>
    </div>
    <div
        className="column"
    >
        <div>
            13
        </div>
    </div>
    <div
        className="column"
    >
        <div>
            14
        </div>
    </div>
</div>,
  <div
    className="row"
>
    <div
        className="column"
    >
        <div>
            15
        </div>
    </div>
    <div
        className="column"
    >
        <div>
            16
        </div>
    </div>
    <div
        className="column"
    >
        <div>
            17
        </div>
    </div>
    <div
        className="column"
    >
        <div>
            18
        </div>
    </div>
    <div
        className="column"
    >
        <div>
            19
        </div>
    </div>
    <div
        className="column"
    >
        <div>
            20
        </div>
    </div>
    <div
        className="column"
    >
        <div>
            21
        </div>
    </div>
</div>,
  <div
    className="row"
>
    <div
        className="column"
    >
        <div>
            22
        </div>
    </div>
    <div
        className="column"
    >
        <div>
            23
        </div>
    </div>
    <div
        className="column"
    >
        <div>
            24
        </div>
    </div>
    <div
        className="column"
    >
        <div>
            25
        </div>
    </div>
    <div
        className="column"
    >
        <div>
            26
        </div>
    </div>
    <div
        className="column"
    >
        <div>
            27
        </div>
    </div>
    <div
        className="column"
    >
        <div>
            28
        </div>
    </div>
</div>,
  <div
    className="row"
>
    <div
        className="column"
    >
        <div>
            29
        </div>
    </div>
    <div
        className="grey column no-day shadow"
    >
        <div />
    </div>
    <div
        className="grey column no-day shadow"
    >
        <div />
    </div>
    <div
        className="grey column no-day shadow"
    >
        <div />
    </div>
    <div
        className="grey column no-day shadow"
    >
        <div />
    </div>
    <div
        className="grey column no-day shadow"
    >
        <div />
    </div>
    <div
        className="grey column no-day shadow"
    >
        <div />
    </div>
</div>,
  <div
    className="row"
/>,
]
`;

exports[`renders Oct 2017 correctly 1`] = `
Array [
  <div
    className="row"
>
    <div
        className="grey column no-day shadow"
    >
        <div />
    </div>
    <div
        className="grey column no-day shadow"
    >
        <div />
    </div>
    <div
        className="grey column no-day shadow"
    >
        <div />
    </div>
    <div
        className="grey column no-day shadow"
    >
        <div />
    </div>
    <div
        className="grey column no-day shadow"
    >
        <div />
    </div>
    <div
        className="grey column no-day shadow"
    >
        <div />
    </div>
    <div
        className="column"
    >
        <div>
            1
        </div>
    </div>
</div>,
  <div
    className="row"
>
    <div
        className="column"
    >
        <div>
            2
        </div>
    </div>
    <div
        className="column"
    >
        <div>
            3
        </div>
    </div>
    <div
        className="column"
    >
        <div>
            4
        </div>
    </div>
    <div
        className="column"
    >
        <div>
            5
        </div>
    </div>
    <div
        className="column"
    >
        <div>
            6
        </div>
    </div>
    <div
        className="column"
    >
        <div>
            7
        </div>
    </div>
    <div
        className="column"
    >
        <div>
            8
        </div>
    </div>
</div>,
  <div
    className="row"
>
    <div
        className="column"
    >
        <div>
            9
        </div>
    </div>
    <div
        className="column"
    >
        <div>
            10
        </div>
    </div>
    <div
        className="column"
    >
        <div>
            11
        </div>
    </div>
    <div
        className="column"
    >
        <div>
            12
        </div>
    </div>
    <div
        className="column"
    >
        <div>
            13
        </div>
    </div>
    <div
        className="column"
    >
        <div>
            14
        </div>
    </div>
    <div
        className="column"
    >
        <div>
            15
        </div>
    </div>
</div>,
  <div
    className="row"
>
    <div
        className="column"
    >
        <div>
            16
        </div>
    </div>
    <div
        className="column"
    >
        <div>
            17
        </div>
    </div>
    <div
        className="column"
    >
        <div>
            18
        </div>
    </div>
    <div
        className="column"
    >
        <div>
            19
        </div>
    </div>
    <div
        className="column"
    >
        <div>
            20
        </div>
    </div>
    <div
        className="column"
    >
        <div>
            21
        </div>
    </div>
    <div
        className="column"
    >
        <div>
            22
        </div>
    </div>
</div>,
  <div
    className="row"
>
    <div
        className="column"
    >
        <div>
            23
        </div>
    </div>
    <div
        className="column"
    >
        <div>
            24
        </div>
    </div>
    <div
        className="column"
    >
        <div>
            25
        </div>
    </div>
    <div
        className="column"
    >
        <div>
            26
        </div>
    </div>
    <div
        className="column"
    >
        <div>
            27
        </div>
    </div>
    <div
        className="column"
    >
        <div>
            28
        </div>
    </div>
    <div
        className="column"
    >
        <div>
            29
        </div>
    </div>
</div>,
  <div
    className="row"
>
    <div
        className="column"
    >
        <div>
            30
        </div>
    </div>
    <div
        className="column"
    >
        <div>
            31
        </div>
    </div>
    <div
        className="grey column no-day shadow"
    >
        <div />
    </div>
    <div
        className="grey column no-day shadow"
    >
        <div />
    </div>
    <div
        className="grey column no-day shadow"
    >
        <div />
    </div>
    <div
        className="grey column no-day shadow"
    >
        <div />
    </div>
    <div
        className="grey column no-day shadow"
    >
        <div />
    </div>
</div>,
]
`;
